src/data/releases/release.service.ts

```diff
@@ -19,57 +19,377 @@
+  try {
+    const release = await releaseModel.getRelease({ _id });
+    if (release) {
+      return toPojo(release);
+    }
+  } catch (e) {
+    logger.error(`Error fetching release by id: ${e}`);
+  }
+
+  return undefined;
+}
+
+/**
+ *
+ * @returns Active release if there is one, otherwise returns undefined
+ */
 export async function getActiveRelease(): Promise<Release | undefined> {
-  const release = await releaseModel.getRelease({ state: ReleaseState.ACTIVE });
+  const release = await releaseModel.getRelease({ state: { $ne: ReleaseState.PUBLISHED } });
   if (release) {
     return toPojo(release);
   }
 
   return undefined;
 }
 
-export async function updateActiveReleaseFiles(files: ReleaseFilesInput): Promise<Release> {
-  const activeRelease = await getActiveRelease();
-  if (activeRelease) {
-    return toPojo(await releaseModel.updateRelease(activeRelease, { files }));
+/**
+ * @returns {Release | undefined} Returns latest release, which is either the active release or the last published release. Returns undefined if no releases have ever been made.
+ */
+export async function getLatestRelease(): Promise<Release | undefined> {
+  const release = await releaseModel.getLatestRelease();
+
+  if (release) {
+    return toPojo(release);
   }
-  return toPojo(await releaseModel.create(files));
+
+  return undefined;
+}
+
+export async function updateActiveReleaseFiles(files: ReleaseFilesInput): Promise<Release> {
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
+  return toPojo(await releaseModel.updateRelease(activeRelease, { files }));
 }
 
 export async function updateActiveReleaseIndices(indices: string[]): Promise<Release> {
-  const activeRelease = await getActiveRelease();
-  if (!activeRelease) {
-    throw new Error('No active release.');
-  }
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
   return toPojo(await releaseModel.updateRelease(activeRelease, { indices }));
 }
 export async function updateActiveReleaseLabel(label: string): Promise<Release> {
-  const activeRelease = await getActiveRelease();
-  if (!activeRelease) {
-    throw new Error('No active release.');
-  }
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
   return toPojo(await releaseModel.updateRelease(activeRelease, { label }));
 }
 export async function updateActiveReleaseSnapshot(snapshot: string): Promise<Release> {
-  const activeRelease = await getActiveRelease();
-  if (!activeRelease) {
-    throw new Error('No active release.');
-  }
+  const activeRelease = await getActiveReleaseOrThrow('No active release.');
   return toPojo(await releaseModel.updateRelease(activeRelease, { snapshot }));
 }
 
-export async function publishActiveRelease(): Promise<Release> {
-  const activeRelease = await getActiveRelease();
+export async function beginCalculatingActiveRelease(): Promise<ReleaseStateChangeResponse> {
```

Comment:
Handling state transitions with a beginCalculating, finishCalculating, beginBuilding... etc.

These return a summary object that reports IF a state transition was performed, a message describing the actions taken or reason no action was taken, the original release state, and a copy of the release after processing.